src/os/bluestore/BlueStore.cc

```diff
@@ -532,6 +532,184 @@ ostream& operator<<(ostream& out, const BlueStore::Buffer& b)
   return out << ")";
 }
 
+// Garbage Collector
+
+void BlueStore::GarbageCollector::process_protrusive_extents(
+  const BlueStore::ExtentMap& extent_map, 
+  uint64_t start_offset,
+  uint64_t end_offset,
+  uint64_t start_touch_offset,
+  uint64_t end_touch_offset,
+  uint64_t min_alloc_size)
+{
+  assert(start_offset <= start_touch_offset && end_offset>= end_touch_offset);
+
+  uint64_t lookup_start_offset = P2ALIGN(start_offset, min_alloc_size);
+  uint64_t lookup_end_offset = ROUND_UP_TO(end_offset, min_alloc_size);
+
+  dout(30) << __func__ << " (hex): [" << std::hex
+           << lookup_start_offset << ", " << lookup_end_offset 
+           << ")" << std::dec << dendl;
+
+  for (auto it = extent_map.seek_lextent(lookup_start_offset);
+       it != extent_map.extent_map.end() &&
+         it->logical_offset < lookup_end_offset;
+       ++it) {
+    uint64_t alloc_unit_start = it->logical_offset / min_alloc_size;
+    uint64_t alloc_unit_end = (it->logical_end() - 1) / min_alloc_size;
+
+    dout(30) << __func__ << " " << *it
+             << "alloc_units: " << alloc_unit_start << ".." << alloc_unit_end
+             << dendl;
+
+    Blob* b = it->blob.get();
+
+    if (it->logical_offset >=start_touch_offset &&
+        it->logical_end() <= end_touch_offset) {
+      // Process extents within the range affected by 
+      // the current write request.
+      // Need to take into account if existing extents
+      // can be merged with them (uncompressed case)
+      if (!b->get_blob().is_compressed()) {
+        if (blob_info_counted && used_alloc_unit == alloc_unit_start) {
+	  --blob_info_counted->expected_allocations; // don't need to allocate
+                                                     // new AU for compressed
+                                                     // data since another
+                                                     // collocated uncompressed
+                                                     // blob already exists
+          dout(30) << __func__  << " --expected:"
+                   << alloc_unit_start << dendl;
+        }
+        used_alloc_unit = alloc_unit_end;
+        blob_info_counted =  nullptr;
+      }
+    } else if (b->get_blob().is_compressed()) {
+
+      // additionally we take compressed blobs that were not impacted
+      // by the write into account too
+      auto b_it =
+        affected_blobs.emplace(b, BlobInfo(b->get_referenced_bytes())).first;
+      BlobInfo& bi = b_it->second;
+      for (auto i = alloc_unit_start; i <= alloc_unit_end; ++i) {
+	if (!used_alloc_unit || i != used_alloc_unit  ) {
+          ++bi.expected_allocations;
+          dout(30) << __func__  << " ++expected_allocations:" << i << dendl;
+	  blob_info_counted =  &bi;
+        }
+	used_alloc_unit = i;
+      }
```

Comment:
I think this loop can be replaced with a

bi.expected_allocations += (alloc_unit_end - alloc_unit_start);
if (used_alloc_unit && used_alloc_unit >= alloc_unit_start && used_alloc_unit < alloc_unit_end) {
  --bi.expected_allocations;
}
used_alloc_unit = alloc_unit_end;
￼
or similar, right?  That avoids looping over AU's in a potentially large blob.